New feature: --skip-path

andrewjhumphrey opened the pull request:
Whilst monkeying about with AWS SSO we discovered it creates roles and
doesn't tag them with anything that we can use to discriminate on with existing
iamy skip functionality.  It does however use a path of `/aws-reserved/`
for all of it's autogenerated roles.

This PR adds the new flag --skip-path and will exclude from consideration
any resource that has a path that starts with any of the specified unwanted 
paths.

This will let us ignore the AWS SSO created roles if we want to.

I've added a test for this that I think is sufficient, but I'm interested
to know if folks think we need to add all of the edge cases to make sure
that strings.HasPrefix is working correctly.

The other question I had was is adding to the isSkippableManagedResource
function the right approach?  Perhaps splitting into a isSkippablePath,
isSkippableTag and isSkippableResourceName function and making the top 
level function just a wrapper would be easier to test deeply?  I'm
not convinced it is worth it, but love some feedback...

andrewjhumphrey committed:
Want to skip things starting with a path we specify

andrewjhumphrey committed:
Add new flag and skip resources with path that matches.

The skipIsManagedResource function was really designed for tags, but
works okay if we pass in the entities path directly to it.

Maybe refactor it into a two functions with single concerns and
wrap them up in an abstraction would be worth it.  I'm not convinced
at this point.

andrewjhumphrey committed:
go fmt fixes

liamdawson commented on iamy.go:
```diff
@@ -44,6 +44,7 @@ func main() {
 		skipCfnTagged   = kingpin.Flag("skip-cfn-tagged", fmt.Sprintf("Shorthand for --skip-tagged %s", cloudformationStackNameTag)).Bool()
 		skipTagged      = kingpin.Flag("skip-tagged", "Skips IAM entities (or buckets associated with bucket policies) tagged with a given tag").Strings()
 		includeTagged   = kingpin.Flag("include-tagged", "Includes IAM entities (or buckets associated with bucket policies) tagged with a given tag").Strings()
+		skipPaths       = kingpin.Flag("skip-path", fmt.Sprintf("Skips IAM entities that have a path starting with the supplied prefixes, repeat flag for multiple prefixes")).Strings()
```
If we want to take this approach (specifically handling prefixes), I'd suggest we try to align with a convention of some kind. I notice `aws iam list-groups` takes a `--path-prefix` rather than `--path`, so perhaps we could use `--skip-path-prefix`?

On the other hand, I'm wondering if we're being unnecessarily specific by adding a flag specifically for excluding prefixes — could we take a more general approach, something like `--exclude '/aws-reserved/*'`?

andrewjhumphrey replied:
Very much prefer the `-prefix` addition to the argument name over `--exclude`, way too much ambiguity in that for my comfort.

andrewjhumphrey replied:
Pushed a change to reflect this a3548ed

andrewjhumphrey committed:
Better naming, skippath -> skippathprefix

We actually skip things based on their path prefix, not their entire path.  Naming this better should reduce confusion later.
It also aligns with at least some of the aws cli (like `iam list-groups`) which names it's path filtering options --path-prefix

liamdawson reviewed (commented):
I'm not really able to review the functionality, but the interface (and intended behaviour) LGTM

patrobinson commented on iamy/aws.go:
```diff
@@ -434,6 +435,14 @@ func (a *AwsFetcher) isSkippableManagedResource(cfnType CfnResourceType, resourc
 		}
 	}
 
+	if len(a.SkipPathPrefixes) > 0 {
```
This if block is unnecessary, since the for loop below will just do nothing if the length is zero.

andrewjhumphrey replied:
Too much C code in my background, fixed this one and the one just above it in 05872ed

patrobinson commented on iamy/aws.go:
```diff
@@ -434,6 +435,14 @@ func (a *AwsFetcher) isSkippableManagedResource(cfnType CfnResourceType, resourc
 	}
 
+	if len(a.SkipPathPrefixes) > 0 {
+		for _, path := range a.SkipPathPrefixes {
```
Can we use this to replace the code below that checks for `aws-service-role`? Given that's doing the same thing

andrewjhumphrey replied:
I didn't want to do that primary for clarity about why things are skipped.  Theoretically we could replace all of the `AWSServiceRole` skipping stuff with a more general purpose --skip-name-pattern or something like that.    I'll tackle that in the next PR.

patrobinson commented:
Should this flag also apply to the pull command?

andrewjhumphrey commented:
> Should this flag also apply to the pull command?

Unless I've pooched something it should do the same thing for both push and pull..

andrewjhumphrey committed:
Remove unnecessary len checks
